test/pkey_name_dup.c

```diff
@@ -0,0 +1,63 @@
+#include <stdlib.h>
+#include <stdio.h>
+
+#include <openssl/bio.h>
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include <openssl/pem.h>
+
+#include "testutil.h"
+
+const char *algs[] = {
+#ifndef OPENSSL_NO_ECX
+    "ED25519",
+#endif
+#ifndef OPENSSL_NO_ML_KEM
+    "ML-KEM-512",
+#endif
+#ifndef OPENSSL_NO_ML_DSA
+    "ML-DSA-44",
+#endif
+    NULL
+};
+
+static int test_name_dup(void)
+{
+    const char **alg;
+    EVP_PKEY *key = NULL;
+    EVP_PKEY_CTX *factory = NULL, *ctx = NULL;
+    int i, ret = 0;
+
+    for (alg = algs; *alg != NULL; ++alg) {
+        EVP_PKEY_CTX_free(factory);
+
+        if (!TEST_ptr(factory = EVP_PKEY_CTX_new_from_name(NULL, *alg, NULL)))
+            goto end;
+
+        for (i = 0; i < 10; ++i) {
+            EVP_PKEY_CTX_free(ctx);
+            EVP_PKEY_free(key);
+            key = NULL;
+            if (!TEST_ptr(ctx = EVP_PKEY_CTX_dup(factory))
+                || !TEST_int_gt(EVP_PKEY_keygen_init(ctx), 0)
+                || !TEST_int_gt(EVP_PKEY_keygen(ctx, &key), 0)) {
+                ERR_print_errors(bio_err);
+                goto end;
+            }
+        }
+    }
+    ret = 1;
+
+ end:
+    EVP_PKEY_CTX_free(factory);
+    EVP_PKEY_CTX_free(ctx);
+    EVP_PKEY_free(key);
+
+    return ret;
+}
+
+int setup_tests(void)
+{
+    ADD_TEST(test_name_dup);
```

Comment:
`ADD_ALL_TESTS(test_name_dup, OSSL_NELEM(algs))` is the more usual pattern for this kind of thing.

Reply:
Yes, but it seemed like overkill, the cases aren't particularly distinct, and not intended to provide comprehensive algorithm coverage, I just wanted to make sure that keygen works for a few *representative* algorithms. So I think of this as a single test. But If others feel that pulling the loop out of the test is *compelling*, I can live with the alternative style.

Reply:
Moved the tests to `evp_test.c` as suggested and used `ADD_ALL_TESTS`.